core/string/fuzzy_search.cpp

```diff
@@ -0,0 +1,348 @@
+		if (substring.y == p_match.token_length) {
+			substring_score += 100;
+		}
+		p_match.score += substring_score;
+	}
+}
+
+void FuzzySearchResult::maybe_apply_score_bonus() {
+	// This adds a small bonus to results which match tokens in the same order they appear in the query.
+	int *token_range_starts = (int *)alloca(sizeof(int) * token_matches.size());
+
+	for (const FuzzyTokenMatch &match : token_matches) {
+		token_range_starts[match.token_idx] = match.interval.x;
+	}
+
+	int last = token_range_starts[0];
+	for (int i = 1; i < token_matches.size(); i++) {
+		if (last > token_range_starts[i]) {
+			return;
+		}
+		last = token_range_starts[i];
+	}
+
+	score += 1;
+}
+
+void FuzzySearchResult::add_token_match(const FuzzyTokenMatch &p_match) {
+	score += p_match.score;
+	match_interval = extend_interval(match_interval, p_match.interval);
+	miss_budget -= p_match.get_miss_count();
+	token_matches.append(p_match);
+}
+
+void remove_low_scores(Vector<FuzzySearchResult> &p_results, const float p_cull_score) {
+	// Removes all results with score < p_cull_score in-place.
+	int i = 0;
+	int j = p_results.size() - 1;
+	FuzzySearchResult *results = p_results.ptrw();
+
+	while (true) {
+		// Advances i to an element to remove and j to an element to keep.
+		while (j >= i && results[j].score < p_cull_score)
+			j--;
+		while (i < j && results[i].score >= p_cull_score)
+			i++;
+		if (i >= j)
+			break;
+		results[i++] = results[j--];
+	}
+
+	p_results.resize(j + 1);
+}
+
+void FuzzySearch::sort_and_filter(Vector<FuzzySearchResult> &p_results) const {
+	if (p_results.is_empty()) {
+		return;
+	}
+
+	float avg_score = 0;
+	float max_score = 0;
+
+	for (const FuzzySearchResult &result : p_results) {
+		avg_score += result.score;
+		max_score = MAX(max_score, result.score);
+	}
+
+	// TODO: Tune scoring and culling here to display fewer subsequence soup matches when good matches
+	//  are available.
+	avg_score /= p_results.size();
+	float cull_score = MIN(cull_cutoff, Math::lerp(avg_score, max_score, cull_factor));
+	remove_low_scores(p_results, cull_score);
+
+	struct FuzzySearchResultComparator {
+		bool operator()(const FuzzySearchResult &A, const FuzzySearchResult &B) const {
+			// Sort on (score, length, alphanumeric) to ensure consistent ordering
+			if (A.score == B.score) {
+				if (A.target.length() == B.target.length()) {
+					return A.target < B.target;
+				}
+				return A.target.length() < B.target.length();
+			}
+			return A.score > B.score;
+		}
+	};
+
+	SortArray<FuzzySearchResult, FuzzySearchResultComparator> sorter;
+
+	if (p_results.size() > max_results) {
+		sorter.partial_sort(0, p_results.size(), max_results, p_results.ptrw());
+		p_results.resize(max_results);
+	} else {
+		sorter.sort(p_results.ptrw(), p_results.size());
+	}
+}
+
+void FuzzySearch::set_query(const String &p_query) {
+	tokens.clear();
+	for (const String &string : p_query.split(" ", false)) {
+		tokens.append({ (int)tokens.size(), string });
```

Comment:
Consider using static_cast<int>() instead of the old c style cast.

Reply:
We use C-style casts more commonly.

Reply:
Quick find shows 9000+ static_casts in the codebase so seems pretty common to me. None the less we shouldn't do things because they're common. static_casts provide type safety. C style casts don't. It's use helps avoid subtle bugs.

Reply:
there are also 293 existing uses of `static_cast<int>` specifically; at first I thought it might be more commonly used in trickier cases. in either case the generated code should be identical, so I don't think it's really worth going back and forth

Reply:
`thirdpart` directory does not count. There are
952 uses of `int()`
1210 uses if `(int)`
49 uses of `static_cast<int>`
It's definitely the least common. But at the same time it's not fully consistent, so whatevs.